[FLINK-19022][runtime]Register the TerminationFuture of ResourceManager and Dispatcher with DispatcherResourceManagerComponent

Tartarus0zm opened the pull request:
## What is the purpose of the change

Help analyze the cause of the abnormality and exit the program normally

## Brief change log

1. add log when RpcEndpoint start/stop failed.
2.Register the TerminationFuture of ResourceManager and Dispatcher with DispatcherResourceManagerComponent.

## Verifying this change

no

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

Tartarus0zm commented:
@tillrohrmann  please take a look, if you have time, thanks

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:
```diff
@@ -88,6 +95,18 @@
+				fatalErrorHandler.onFatalError(throwable);
+			}
+		};
+		dispatcherRunner.getTerminationFuture().whenComplete(terminateAction);
```
I think the `DispatcherRunner.getTerminationFuture` won't complete if the `Dispatcher` terminates. The problem is that we only forward the result of the `DispatcherLeaderProcess.terminationFuture` to the `DispatcherRunner.terminationFuture` if we call `closeAsync`. I would suggest to add tests to verify the intended behavior.

Tartarus0zm replied:
But how we get the Dispatcher's TerminationFuture in DispatcherResourceManagerComponent?
Please give a hint, thanks

Tartarus0zm replied:
@tillrohrmann  I need some help,  I took a look at this piece of code,  `DispatcherRunner.terminationFuture` is not connected to `Dispatcher`, so how we get the `Dispatcher.TerminationFuture` in `DispatcherResourceManagerComponent`?
Please give a hint, thanks

rmetzger replied:
@tillrohrmann is on vacation for a few weeks starting from today. He won't be able to resolve your questions.

I would suggest to follow his suggestions and add test cases reproducing the problem first. This should make it easier finding the right location in the code in the code handle errors.

Tartarus0zm replied:
@rmetzger  thanks, I try to solve it myself

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunner.java:
```diff
@@ -107,13 +112,18 @@ public void grantLeadership(UUID leaderSessionID) {
+			FutureUtils.assertNoException(
+				previousDispatcherLeaderProcessTerminationFuture.thenRun(newDispatcherLeaderProcess::start));
+		} catch (Throwable t) {
+			terminationFuture.completeExceptionally(t);
+			throw t;
```
Why is this needed?

tillrohrmann reviewed (changes requested):
Thanks for creating this PR @Tartarus0zm. I think the current implementation still has a couple of problems. The main part is the forwarding and usage of the termination future of the `DispatcherRunner`.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:
```diff
@@ -74,22 +79,46 @@
 
+	private void failOnPrematureTermination(CompletableFuture<DispatcherGateway> dispatcherGatewayCompletableFuture) {
+		dispatcherGatewayCompletableFuture.whenComplete((dispatcher, throwable) -> {
+			if (dispatcher != null && dispatcher instanceof Dispatcher) {
```
This breaks the DispatcherGateway abstraction and reinforces the assumption that the RM and Dispatcher run in the same JVM which may change in the future.
Why can we not use the shutdown future of the `DispatcherRunner`?

Tartarus0zm replied:
The shutdown future of the `DispatcherRunner` has been registed with the shutDown Future of `DispatcherResourceManagerComponent`.  I think there should be no need to deal with it again.
How about that only deals with the Termination Future of ResourceManager here?

zentol replied:
Are the DispatcherRunner and Dispatcher termination futures linked in any way? As in, if the Dispatcher shuts down _at any point_, then it is guaranteed that the DispatcherRunner termination future is also being completed?

Tartarus0zm replied:
if we use the shutdown future of the DispatcherRunner like this
```
private void failOnPrematureTermination() {
	CompletableFuture.anyOf(dispatcherRunner.getShutDownFuture(), resourceManager.getTerminationFuture())
		.whenComplete((ignored, t) -> {
			if (t == null) {
				LOG.error("Dispatcher/ResourceManager shut down because something unexpected happen!");
			}
			if (isRunning.get()) {
				LOG.warn("DispatcherResourceManagerComponent shut down because the Dispatcher/ResourceManager unexpected terminated.");
				fatalErrorHandler.onFatalError(t);
			}
		});
}
```
in per-job mode, `ClusterEntrypoint.shutDownAsync()` may not have a chance to execute, because `ClusterEntrypoint.onFatalError` will call `System.exit(RUNTIME_FAILURE_RETURN_CODE);`
Do you have any good suggestions for revision?

Tartarus0zm replied:
We first execute `failOnPrematureTermination();` and then execute `registerShutDownFuture();` 
How about this?  @zentol

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:
```diff
@@ -74,22 +79,46 @@
+						}
+					});
+			} else {
+				LOG.warn("DispatcherGateway({}) not instance of Dispatcher or failed to get DispatcherGateway!", dispatcher, throwable);
```
This kind of implementation details should not be written to the logs.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:
```diff
@@ -74,22 +79,46 @@
+				CompletableFuture.anyOf(((Dispatcher) dispatcher).getTerminationFuture(), resourceManager.getTerminationFuture())
+					.whenComplete((ignored, t) -> {
+						if (isRunning.get()) {
+							fatalErrorHandler.onFatalError(t);
```
It would be good to log a warning here that the DRMC shut down because the dispatcher/rm unexpected terminated.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:
```diff
@@ -74,22 +79,46 @@
+		dispatcherGatewayCompletableFuture.whenComplete((dispatcher, throwable) -> {
+			if (dispatcher != null && dispatcher instanceof Dispatcher) {
+				CompletableFuture.anyOf(((Dispatcher) dispatcher).getTerminationFuture(), resourceManager.getTerminationFuture())
+					.whenComplete((ignored, t) -> {
```
It may be a good idea to check whether t is indeed not null and log something accordingly if this is not the case. If it is null, then the process shut down without an error unexpectedly, which if I understand things correctly should not happen.

Tartarus0zm replied:
We has modify `Dispatcher` and `ResourceManager` catch Exception to catch Throwable ，and then call `onFatalError` , so I think `DispatcherResourceManagerComponent#failOnPrematureTermination` is not very necessary, but @tillrohrmann  thinks it is safer

zentol replied:
I don't see how that relates to my comment. We will call the fatal error handler either way; the issue is that if t == null no reason for the shutdown will be logged-

Tartarus0zm replied:
sorry, my understanding is wrong.

tillrohrmann committed:
[FLINK-19022][rpc] Log if RpcEndpoint failed its termination

In order to not swallow the failure reason of failed RpcEndpoints, we are now
logging the cause of a RpcEndpoint failure.

tillrohrmann committed:
[FLINK-19022] Let TaskExecutor, Dispatcher, RM fail fatally if onStart throws a Throwable

If TaskExecutor., Dispatcher. or ResourceManager.onStart throws a Throwable, the system is
no longer in a valid state. Hence, this commit ensures that the components will properly
call FatalErrorHandler.onFatalError in such a case.

tillrohrmann committed:
[hotfix] Let DispatcherResourceManagerComponent accept AutoCloseableAsync instead of WebMonitorEndpoint

This commit improves the testability of DispatcherResourceManagerComponent by allowing to pass
in a AutoCloseableAsync implementation.

tillrohrmann committed:
[hotfix] Introduce ResourceManagerService to make DispatcherResourceManagerComponent easier to test

By passing an interface instead of ResourceManager, it is easier to test the DispatcherResourceManagerComponent.

tillrohrmann committed:
[FLINK-19022] Fail DispatcherResourceManagerComponent fatally if ResourceManager terminates unexpectedly

An unexpected termination of the ResourceManager indicates an invalid state. Hence, the
DispatcherResourceManagerComponent should react to it by calling the FatalErrorHandler.

tillrohrmann committed:
[hotfix] Add FlinkMatchers.willNotComplete to assert that a future times out

tillrohrmann committed:
[hotfix] Correct TestingDispatcherGatewayService closeAsync behavior

Let TestingDispatcherGatewayService complete the termination future when calling closeAsync.

tillrohrmann committed:
[FLINK-19022] Fail fatally if DispatcherGatewayService terminates unexpectedly in AbstractDispatcherLeaderProcess

This commit ensures that we fail fatally if the DispatcherGatewayService termiantes unexpectedly in the
AbstractDispatcherLeaderProcess. This ensures that the system won't stay in an invalid state.

tillrohrmann committed:
[FLINK-19022] Fail TaskManagerRunner fatally if TaskExecutor terminates unexpectedly

An unexpected termination of the TaskExecutor indicates that the system has reached an
invalid state. Therefore, the runner will fail fatally if this happens.

tillrohrmann commented:
Sorry for the late response @Tartarus0zm. I have take a look at your solution and reworked it a bit. Please take a look at the pushed commits.

@zentol I also appreciate your review of the changes.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/AbstractDispatcherLeaderProcess.java:
```diff
@@ -139,6 +139,8 @@ public final UUID getLeaderSessionId() {
 	private void closeInternal() {
 		log.info("Stopping {}.", getClass().getSimpleName());
 
+		state = State.STOPPED;
```
Does this change belong into `cd14c5907bdbd88861f9e74448ded52d5c15a7a6`? (Because it requires other components to be stopped after the leader process transitioned to STOPPED.

tillrohrmann replied:
Yes I think it is correct. `State.STOPPED` effectively means that `closeInternal` has been called. But it does not say that the stopping is complete.

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponentTest.java:
```diff
@@ -0,0 +1,95 @@
+			TestingFatalErrorHandler fatalErrorHandler,
+			TestingResourceManagerService resourceManagerService) {
+		return new DispatcherResourceManagerComponent(
+			TestingDispatcherRunner.newBuilder().build(),
+				resourceManagerService,
+			new SettableLeaderRetrievalService(),
+			new SettableLeaderRetrievalService(),
+				FutureUtils::completedVoidFuture,
+				fatalErrorHandler);
```
odd indentation

tillrohrmann replied:
You are right. Somehow IntelliJ uses a different indentation nowadays... Should check the configuration.

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunnerTest.java:
```diff
@@ -144,6 +148,58 @@ public void testGenerateTaskManagerResourceIDWithLocalRpcService() throws Except
+				.build();
+		final TaskManagerRunner taskManagerRunner = createTaskManagerRunner(
+				createConfiguration(),
+				(configuration,
+				resourceID,
+				rpcService,
+				highAvailabilityServices,
+				heartbeatServices,
+				metricRegistry,
+				blobCacheService,
+				localCommunicationOnly,
+				externalResourceInfoProvider,
+				fatalErrorHandler) -> taskExecutorService);
```
maybe introduce a utility method that returns this factory; it is easy to miss that the list of arguments is part of the factory, and not the constructor.

tillrohrmann replied:
Good point. I will update it.

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunnerTest.java:
```diff
@@ -144,6 +148,58 @@ public void testGenerateTaskManagerResourceIDWithLocalRpcService() throws Except
+				.build();
+		final TaskManagerRunner taskManagerRunner = createTaskManagerRunner(
+				createConfiguration(),
+				(configuration,
+				resourceID,
+				rpcService,
+				highAvailabilityServices,
+				heartbeatServices,
+				metricRegistry,
+				blobCacheService,
+				localCommunicationOnly,
+				externalResourceInfoProvider,
+				fatalErrorHandler) -> taskExecutorService);
```
same as above

tillrohrmann replied:
Good idea. Will do.

zentol reviewed (approved):
+1

tillrohrmann committed:
fixup! [FLINK-19022] Fail DispatcherResourceManagerComponent fatally if ResourceManager terminates unexpectedly

tillrohrmann committed:
fixup! [FLINK-19022] Fail TaskManagerRunner fatally if TaskExecutor terminates unexpectedly

tillrohrmann committed:
fixup! [FLINK-19022] Fail DispatcherResourceManagerComponent fatally if ResourceManager terminates unexpectedly

tillrohrmann committed:
fixup! [FLINK-19022] Fail DispatcherResourceManagerComponent fatally if ResourceManager terminates unexpectedly

tillrohrmann committed:
fixup! [FLINK-19022] Fail TaskManagerRunner fatally if TaskExecutor terminates unexpectedly

tillrohrmann committed:
fixup! [FLINK-19022] Fail TaskManagerRunner fatally if TaskExecutor terminates unexpectedly

tillrohrmann commented:
Thanks for the review @zentol. Merging this PR now.
